src/SmartRange.ts

```diff
@@ -0,0 +1,207 @@
+ * r.length // 5
+ * const a = [...r]; // [0,2,4,6,8]
+ * // stepsize defaults to 1 | -1
+ * new SmartRange(2,-3) // 2, 1, 0, -1, -2
+ * ```
+ * @implements {Iterable<number>}
+ * @implements {Iterator<number, undefined>}
+ * @implements {ArrayLike<number>}
+ * @author Leo Schurrer <l.schurrer@outlook.com>
+ * @license MIT
+ */
+export default class SmartRange
+    implements ArrayLike<number>, Iterator<number, undefined>, Iterable<number>
+{
+    // index signature only usefull if class is used with Proxy
+    [index: number]: number;
+
+    // private members
+    #start: number;
+    #end: number;
+    #step: number;
+    /** number of steps performed by the Iterator protocol */
+    #doneSteps: number = 0;
+
+    /**
+     * ## Constructor
+     * values need to be integers
+     * @param start start of the range can be larger than end with negative step
+     * @param end end of the range
+     * @param step step between values in the range - can be negative defaults to 1 or -1 if end < start
+     */
+    constructor(start: number, end: number, step?: number) {
+        //  TODO accept options for excludeStart, includeEnd
+        if (
+            !Number.isInteger(start) || !Number.isInteger(end) || step
+                ? !Number.isInteger(step)
+                : false
+        ) {
+            throw new TypeError(
+                "SmartRange: start, end and step must be integers"
+            );
+        }
+        this.#start = start;
+        this.#end = end;
+        this.#step =
+            step === undefined || step === 0 ? this.#defaultStep : step;
+    }
+
+    get start(): number {
+        return this.#start;
+    }
+    get end(): number {
+        return this.#end;
+    }
+    get step(): number {
+        return this.#step;
+    }
+    get length(): number {
+        if (this.#isInvalid) return 0;
+        const res = Math.abs(this.#end - this.#start) / this.#step;
+        return res < 0 ? Math.floor(res) : Math.ceil(res);
+    }
+    get #defaultStep(): number {
+        return this.#start < this.#end ? 1 : -1;
+    }
+    get #isInvalid(): boolean {
+        const cond1 = this.#start < this.#end && this.#step < 0;
+        const cond2 = this.#end < this.#start && 0 < this.#step;
+        return cond1 || cond2;
+    }
+
+    set start(v: number) {
+        if (this.#start === v) return;
+        if (!Number.isInteger(v))
+            throw new TypeError("SmartRange: start needs to be integer");
+        this.#start = v;
+    }
+    set end(v: number) {
+        if (this.#end === v) return;
+        if (!Number.isInteger(v))
+            throw new TypeError("SmartRange: end needs to be integer");
+        this.#end = v;
+    }
+    set step(v: number | undefined) {
+        if (this.step === v) return;
+        if (!Number.isInteger(v))
+            throw new TypeError("SmartRange: step needs to be integer");
+        // FIXME what happens to this.#doneSteps when step size is changed after steps are already done
```

Comment:
FIXME. Should this be in the first version?

Reply:
I'll changed prio to TODO. For now I still have no good resolution, but at least it's mentioned in the documentation.